editor/src/core/tailwind/tailwind-compilation.ts

```diff
@@ -0,0 +1,110 @@
+import type { ProjectContentTreeRoot } from 'utopia-shared/src/types'
+import { getProjectFileByFilePath, walkContentsTree } from '../../components/assets'
+import { interactionSessionIsActive } from '../../components/canvas/canvas-strategies/interaction-state'
+import { CanvasContainerID } from '../../components/canvas/canvas-types'
+import {
+  Substores,
+  useEditorState,
+  useRefEditorState,
+} from '../../components/editor/store/store-hook'
+import { importDefault } from '../es-modules/commonjs-interop'
+import { rescopeCSSToTargetCanvasOnly } from '../shared/css-utils'
+import type { RequireFn } from '../shared/npm-dependency-types'
+import { TailwindConfigPath } from './tailwind-config'
+import { memoize } from '../shared/memoize'
+import { equalityFnForProjectContents } from '../shared/project-contents-utils'
+
+const TAILWIND_INSTANCE: { current: Tailwindcss | null } = { current: null }
+
+export function isTailwindEnabled(): boolean {
+  return TAILWIND_INSTANCE.current != null
+}
+
+function ensureElementExists({ type, id }: { type: string; id: string }) {
+  let tag = document.getElementById(id)
+  if (tag == null) {
+    tag = document.createElement(type)
+    tag.id = id
+    document.head.appendChild(tag)
+  }
+  return tag
+}
+
+async function generateTailwindStyles(tailwindCss: Tailwindcss, allCSSFiles: string) {
+  const contentElement = document.getElementById(CanvasContainerID)
+  if (contentElement == null) {
+    return
+  }
+  const content = contentElement.outerHTML
+  const styleString = await tailwindCss.generateStylesFromContent(allCSSFiles, [content])
+  const style = ensureElementExists({ type: 'style', id: 'utopia-tailwind-jit-styles' })
+  style.textContent = rescopeCSSToTargetCanvasOnly(styleString)
+}
+
+function getCssFilesFromProjectContents(projectContents: ProjectContentTreeRoot) {
+  let files: string[] = []
+  walkContentsTree(projectContents, (path, file) => {
+    if (file.type === 'TEXT_FILE' && path.endsWith('.css')) {
+      files.push(file.fileContents.code)
+    }
+  })
+  return files
+}
+
+function generateTailwindClassesInner(
+  projectContents: ProjectContentTreeRoot,
+  requireFn: RequireFn,
+) {
+  const allCSSFiles = getCssFilesFromProjectContents(projectContents).join('\n')
+  const rawConfig = importDefault(requireFn('/', TailwindConfigPath))
+  const tailwindCss = createTailwindcss({ tailwindConfig: rawConfig as TailwindConfig })
+  TAILWIND_INSTANCE.current = tailwindCss
+  void generateTailwindStyles(tailwindCss, allCSSFiles)
+}
+
+const generateTailwindClasses = memoize(generateTailwindClassesInner, {
+  maxSize: 1,
+  matchesArg: equalityFnForProjectContents,
+})
+
+export const useTailwindCompilation = (requireFn: RequireFn) => {
+  const projectContents = useEditorState(
+    Substores.projectContents,
+    (store) => store.editor.projectContents,
+    'useTailwindCompilation projectContents',
+  )
+
+  const isInteractionActiveRef = useRefEditorState((store) =>
```

Comment:
I think a safer proxy is ElementsToRerenderGLOBAL.current, but we should definitely have an official function for this

Reply:
https://github.com/concrete-utopia/utopia/pull/6401/commits/ef7ec9c60bf04122b82ff35739183678928e460e